### PR TITLE
fix: initialization HTTP clients with merged `DevWorkspaceOperatorConfig`

### PR DESCRIPTION
**Pull request by tolusha**
### What does this PR do?
This PR changes the HTTP client initialization to use the merged DevWorkspaceOperatorConfig instead of only the global config.

### What issues does this PR fix or reference?
https://github.com/eclipse-che/che/issues/23870
https://github.com/eclipse-che/che-operator/pull/2137

### Is it tested? How?

Followed https://github.com/eclipse-che/che-operator/pull/2137

### PR Checklist

- [x] E2E tests pass (when PR is ready, comment `/test v8-devworkspace-operator-e2e, v8-che-happy-path` to trigger)
    - [x] `v8-devworkspace-operator-e2e`: DevWorkspace e2e test
    - [x] `v8-che-happy-path`: Happy path for verification integration with Che

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved health check reliability with proper response body cleanup.

* **Tests**
  * Added comprehensive unit tests for HTTP client creation, caching, and concurrent access.

* **Chores**
  * Refactored HTTP client initialization to use a thread-safe factory pattern with improved routing configuration support.

### CODE REVIEW

**Comment by tolusha** [comment]
Hi! I'm **che-ai-assistant** — I help with your pull requests.

**Available commands**:
- `/che-ai-assistant generate-che-doc` — Generate a documentation PR based on this PR's changes
- `/che-ai-assistant ok-pr-review` — Run a comprehensive PR review (summary, code review, deep review, impact analysis)
- `/che-ai-assistant help` — Show this help message

**Comment by tolusha** [comment]
# PR Review: Fix initialization HTTP clients with merged DevWorkspaceOperatorConfig

## 📋 Summary

This PR refactors HTTP client initialization in the DevWorkspace Operator to use a thread-safe factory pattern that properly merges DevWorkspaceOperatorConfig settings. The changes address an issue where HTTP clients were only initialized with global config, missing workspace-specific routing configurations.

**Key Changes:**
- Introduced `HttpClientsFactory` interface and `DefaultHttpClientsFactory` implementation
- Added client caching with invalidation based on configuration changes
- Implemented proper mutex-based concurrency control
- Added comprehensive unit tests (219 new lines)
- Fixed resource leak by adding `resp.Body.Close()` in health check

**Files Changed:** 4 files, +447/-67 lines
- `controllers/workspace/devworkspace_controller.go` (+6/-4)
- `controllers/workspace/http.go` (+215/-58) 
- `controllers/workspace/http_test.go` (+219/-4)
- `controllers/workspace/status.go` (+7/-1)

---

## 🔍 Code Review

### ✅ Strengths

1. **Solid Architecture**: The factory pattern is well-suited for this use case, enabling:
   - Centralized HTTP client creation logic
   - Configuration-based caching
   - Easy testing via interface abstraction

2. **Thread Safety**: Proper use of `sync.RWMutex` with read-write lock pattern:
   - Fast path: RLock for cache hits
   - Double-checked locking to prevent race conditions
   - Write lock only when rebuilding clients

3. **Comprehensive Testing**: Excellent test coverage including:
   - Concurrent access safety tests
   - Cache invalidation scenarios
   - Proxy and certificate configuration changes
   - Mock factory for existing tests

4. **Resource Management**: Fixed response body leak in `checkServerStatus` with proper `defer resp.Body.Close()`

5. **Configuration Merging**: Now correctly uses workspace-specific routing config instead of just global config

### ⚠️ Issues & Concerns

#### **Critical Issues**

1. **Nil Pointer Risk in `GetHttpClient`** (Line 84-86)
   ```go
   httpClient := httpClientsFactory.GetHttpClient(ctx, config.Routing)
   ```
   If `httpClientsFactory` is nil (e.g., `SetupHttpClientsFactory` failed silently or wasn't called), this will panic. The error from `SetupWithManager` should be checked, but there's no guarantee.

   **Recommendation**: Add nil check or ensure factory is always initialized.

2. **Error Handling in Certificate Loading** (Line 238-241)
   ```go
   if !caCertPool.AppendCertsFromPEM([]byte(certsPem)) {
       h.logger.Error(fmt.Errorf("failed to parse..."), "Could not append CA certificates to pool")
   }
   ```
   The error is logged but the invalid cert data is silently ignored. This could lead to unexpected TLS failures without clear indication of the root cause.

   **Recommendation**: Consider returning the partially-configured pool with a warning, or track parsing failures.

#### **Medium Priority Issues**

3. **System Cert Pool Clone Cost**
   ```go
   caCertPool := h.systemCertPool.Clone()
   ```
   Cloning the system cert pool on every client creation may have performance implications if the pool is large. Since `systemCertPool` is immutable after init, a single shared pool could work.

   **Recommendation**: Benchmark if this becomes a bottleneck; consider reusing base pool.

4. **ConfigMap Read on Every Client Request**
   The `readCertificates` method reads the ConfigMap from the API server on every `GetHttpClient` call (even cache hits use it for comparison).

   **Recommendation**: Consider caching the ConfigMap read result or using a watch/informer pattern.

5. **Timeout Inconsistency**
   - Regular HTTP client: `10 * time.Second`
   - Health check client: `500 * time.Millisecond`
   
   The regular client timeout seems high for cluster-internal operations. Is 10s intentional?

   **Recommendation**: Document timeout rationale or make configurable.

#### **Minor Issues**

6. **Missing Context Propagation**
   The `readCertificates` method uses `ctx` from the caller, but `SetupHttpClientsFactory` and initial setup don't use contexts. Consider consistent context handling.

7. **Copyright Year Update**
   Updated to 2026 but current date context shows 2026-06-17. Verify this is intentional.

---

## 🔬 Deep Review

### **Architectural Analysis**

**Pattern Choice**: ✅ Factory pattern is appropriate
- Centralizes client creation logic
- Enables dependency injection for testing
- Supports configuration-based caching

**Concurrency Model**: ✅ Well-implemented
- Double-checked locking prevents races
- RWMutex allows concurrent reads
- Appropriate lock granularity

**Error Handling**: ⚠️ Needs improvement
- Silent failures in cert parsing could mask issues
- No circuit breaker for repeated ConfigMap read failures
- Nil factory panic risk

### **Security Considerations**

1. ✅ **TLS Verification**: Regular client properly verifies certificates
2. ✅ **Health Check Isolation**: Separate client for `InsecureSkipVerify` is correct - limited to internal health checks
3. ⚠️ **Certificate Validation**: Invalid PEM data is logged but not surfaced - could hide misconfigurations
4. ✅ **Proxy Support**: Properly handles proxy configuration with NO_PROXY support

### **Performance Analysis**

**Caching Strategy**: ✅ Efficient
- Clients cached until config changes
- Fast path (RLock) for cache hits
- Minimal allocations in hot path

**Potential Bottlenecks**:
1. ⚠️ API calls to read ConfigMap on every `GetHttpClient` call
2. ⚠️ System cert pool cloning
3. ⚠️ Deep equality comparisons on every call

**Recommendations**:
- Add metrics for cache hit/miss rates
- Consider ConfigMap watch instead of repeated reads
- Benchmark under high reconciliation load

### **Testing Quality**

✅ **Excellent test coverage**:
- Concurrent access safety
- Cache invalidation logic
- Configuration change detection
- Mock implementation for existing tests

⚠️ **Missing test cases**:
- Error scenarios (invalid certs, missing ConfigMap)
- Nil routing config handling
- Proxy configuration edge cases
- System cert pool loading failure

---

## 📊 Impact Assessment

### **Functional Impact**

**Positive Changes**:
1. ✅ Fixes the core issue: HTTP clients now respect workspace-specific routing config
2. ✅ Improved concurrency safety for multi-workspace environments
3. ✅ Fixed resource leak (response body not closed)
4. ✅ Better separation of concerns (factory pattern)

**Risk Areas**:
1. ⚠️ Breaking change in global variable usage (`httpClient` → `httpClientsFactory`)
2. ⚠️ Different initialization timing could expose race conditions
3. ⚠️ Performance impact from ConfigMap reads needs monitoring

### **Compatibility**

**Backward Compatibility**: ✅ Maintained
- Test helper `SetupHttpClientsForTesting` preserves existing test interface
- Public API unchanged

**Upgrade Path**: ✅ Smooth
- No migration required
- Changes are internal to the controller

### **Operational Impact**

**Performance**: ⚠️ Needs monitoring
- ConfigMap reads on every reconciliation loop
- Client recreation when configs change
- Increased memory from caching and mutexes

**Observability**: ⚠️ Limited
- No metrics for cache efficiency
- Errors logged but not counted
- No visibility into client recreation frequency

**Recommendations**:
1. Add Prometheus metrics: `http_client_cache_hits`, `http_client_cache_misses`, `http_client_recreation_total`
2. Add structured logging for config changes
3. Consider health check endpoint to verify factory state

### **Deployment Considerations**

**Rollout Risk**: 🟢 Low
- Changes are internal
- Comprehensive testing included
- E2E tests passed

**Rollback**: 🟢 Easy
- Simple revert if issues found
- No database/state changes

---

## 💡 Recommendations

### **High Priority**

1. **Add nil factory safeguard**:
   ```go
   if httpClientsFactory == nil {
       return reconcile.Result{}, fmt.Errorf("http clients factory not initialized")
   }
   ```

2. **Improve cert parsing error visibility**:
   - Return error count from `getCaCertPool`
   - Add metric for parsing failures
   - Consider failing fast if all certs are invalid

3. **Optimize ConfigMap reads**:
   - Cache ConfigMap reads with TTL
   - Or use informer/watch pattern
   - Only re-read when ResourceVersion changes

### **Medium Priority**

4. **Add observability**:
   - Metrics for cache hit/miss
   - Client recreation counters
   - ConfigMap read errors

5. **Document timeout values**:
   - Why 10s for regular client?
   - Why 500ms for health checks?

6. **Add integration test**:
   - Verify workspace-specific config is actually used
   - Test config updates trigger client rebuild

### **Low Priority**

7. **Consider configuration validation**:
   - Validate proxy URLs at creation time
   - Validate cert PEM format before caching

8. **Add comments for thread safety**:
   - Document why double-checked locking is needed
   - Explain cache invalidation logic

---

## ✅ Approval Recommendation

**Status**: ✅ **APPROVE with minor suggestions**

This PR successfully addresses the stated issue and introduces a well-architected solution. The factory pattern is appropriate, thread safety is properly implemented, and test coverage is excellent.

**Blockers**: None - the identified issues are primarily optimizations and observability enhancements that can be addressed in follow-up PRs.

**Confidence Level**: High
- E2E tests passed
- Comprehensive unit tests
- Well-structured refactoring
- Addresses root cause of the issue

**Suggested Follow-ups**:
1. Add metrics for cache efficiency monitoring
2. Optimize ConfigMap reads (watch pattern or caching)
3. Add integration test for workspace-specific config
4. Document timeout rationale

---

## 🎯 Summary

**What this PR does well**:
- Solves the core problem (merged config usage)
- Excellent test coverage
- Thread-safe implementation
- Fixes resource leak

**What could be improved**:
- ConfigMap read performance
- Error observability
- Nil factory safeguards

**Overall**: A solid refactoring that improves the codebase. The suggested improvements are optimizations rather than blockers. Great work! 🚀

---

*Review generated by Claude Code PR Review on 2026-06-17*

**Review by dkwon17 — APPROVED** [review]

File: controllers/workspace/http.go [hunk 1]
```diff
@@ -32,87 +34,242 @@ import (
...
+func (h *DefaultHttpClientsFactory) createHttpClient(routingConfig *controller.RoutingConfig, certsCM *corev1.ConfigMap) *http.Client {
+	transport := http.DefaultTransport.(*http.Transport).Clone()
+	transport.Proxy = h.getProxyFunc(routingConfig)
+	transport.TLSClientConfig = &tls.Config{
+		RootCAs: h.getCaCertPool(certsCM),
+	}
+
+	return &http.Client{
 		Transport: transport,
+		Timeout:   10 * time.Second,
```
Comment (rohanKanojia) [inline controllers/workspace/http.go:132]: Why `10` seconds specifically? The health-check client had `500ms`? Is this based on existing routing behavior or expected response times? I assume 10s is new for the devfile client.

Comment (tolusha) [reply controllers/workspace/http.go:132]: it can be 3/5/10.
I don't know what it the best value, but we need something

**Comment by rohanKanojia** [comment]
I tested the following steps mentioned in https://github.com/eclipse-che/che-operator/pull/2137 and can confirm it's working as expected ✔️

**Review by rohanKanojia — APPROVED** [review]
